wurst/math/Quaternions.wurst

```diff
@@ -0,0 +1,215 @@
+
+public function quat.op_mult(real scalar) returns quat
+    return quat(this.x * scalar, this.y * scalar, this.z * scalar, this.w * scalar)
+
+public function real.op_mult(quat q) returns quat
+    return q * this
+
+public function quat.dot(quat q) returns real
+    return this.x * q.x + this.y * q.y + this.z * q.z + this.w * q.w
+
+/** Represents the composition of two rotations. */
+public function quat.cross(quat q) returns quat
+    return quat(this.x * q.w + this.y * q.z - this.z * q.y + this.w * q.x,
+                -this.x * q.z + this.y * q.w + this.z * q.x + this.w * q.y,
+                this.x * q.y - this.y * q.x + this.z * q.w + this.w * q.z,
+                -this.x * q.x - this.y * q.y - this.z * q.z + this.w * q.w)
+
+/** Represents the same rotation in opposite direction. */
+public function quat.conj() returns quat
+    return quat(-this.x, -this.y, -this.z, this.w)
+
+public function quat.length() returns real
+    return SquareRoot(this.x.squared() + this.y.squared() + this.z.squared() + this.w.squared())
+
+/** Returns zero quaternion if lenght is zero. */
+public function quat.norm() returns quat
+    let length = this.length()
+    return length > 0.002 ? quat(this.x / length, this.y / length, this.z / length, this.w / length) : ZEROQ
+
+public function quat.toMatrix() returns mat3
+    let q = this.norm()
+    let x = q.x
+    let y = q.y
+    let z = q.z
+    let w = q.w
+    return mat3(1 - 2 * y * y - 2 * z * z,      2 * x * y - 2 * z * w,      2 * x * z + 2 * y * w,
+                2 * x * y + 2 * z * w,          1 - 2 * x * x - 2 * z * z,  2 * y * z - 2 * x * w,
+                2 * x * z - 2 * y * w,          2 * y * z + 2 * x * w,      1 - 2 * x * x - 2 * y * y)
+
+/** Extracts X-axis from a quaternion that represents rotation. */
+public function quat.getX() returns vec3
+    return vec3(1 - 2 * this.y * this.y - 2 * this.z * this.z,
+                2 * this.x * this.y + 2 * this.z *this. w,
+                2 * this.x * this.z - 2 * this.y * this.w)
+
+/** Extracts Y-axis as a vector from a quaternion that represents rotation. */
+public function quat.getY() returns vec3
+    return vec3(2 * this.x * this.y - 2 * this.z * this.w,
+                1 - 2 * this.x * this.x - 2 * this.z * this.z,
+                2 * this.y * this.z + 2 * this.x * this.w)
+
+/** Extracts Z-axis as a vector from a quaternion that represents rotation. */
+public function quat.getZ() returns vec3
+    return vec3(2 * this.x * this.z + 2 * this.y * this.w,
+                2 * this.y * this.z - 2 * this.x * this.w,
+                1 - 2 * this.x * this.x - 2 * this.y * this.y)
+                
+/** Linear intERPolation.
+q - targeted quaternion. p - progress from 0 to 1. */
+public function quat.lerp(quat q, real p) returns quat
+    quat result
+    if this.dot(q).abs() < 0.002
+        result = q
+    else
+        result = this * (1 - p) + q * p
+    return result
+
+/** Linear intERPolation.

```

Comment:
Is this a joke about ERP? lol

Reply:
yea but why the casing

Reply:
Because I cannot show it otherwise?

Reply:
show what? the word is "interpolation", you write "intERPolation", you link some slerp/nlerp stuff which seems totally unrelated?